### DISCUSSION TITLE
Upgrade mechanism

### DISCUSSION THREAD

**Discussion by Pyjou** [jeremylvln/Shulker#354] [Q&A] [answered]
I have a big question about the upgrading of a server with Shulker.

In my case, let's say I want to upgrade a plugin on one of my server fleets (proxy or not). I change the value of my url in env vars, which will be downloaded when my server starts up, and send it to kube. But we agree that kube will automatically kill all servers to apply the change, and automatically kick players. That's not what we'd generally want.

Do you think there would be a mechanism to wait until the end of a game session (like in a mini-game) before launching the upgrade? Or simply not lauchn the upgrade on servers already running and simply apply it to new servers?

What do you think about this subject?

**Comment by jeremylvln** [comment]
Hi,

Agones has some nice documentation about this specific use case. You have to tell Agones that a server is `Allocated` so it does consider it "in use" and does not touch it until (1) it is stopped or (2) it is marked again as `Ready`.

Shulker does automatically the job of setting the `GameServer` as `Ready` thanks to its agent plugin. But if you want to _reserve_ it, you'll have to do it yourself.

Agones [provides an SDK](https://agones.dev/site/docs/guides/client-sdks/#lifecycle-management) with appropriate methods. They do not however give a Java implementation. I created one for Shulker that you can use directly:
- https://github.com/jeremylvln/Shulker/blob/main/packages/google-agones-sdk/bindings/java/src/main/java/com/agones/dev/sdk/AgonesSDK.java#L13
- https://github.com/jeremylvln/Shulker/blob/main/packages/google-agones-sdk/bindings/java/src/main/java/com/agones/dev/sdk/AgonesSDKImpl.java#L126

It does not contain any Javadoc for now, I was not expecting someone to use it, but you absolutely can. I use it in a large Shulker cluster :) It is published in [my personal Maven here](https://maven.jeremylvln.fr/ui/repos/tree/General/shulker-releases/io/shulkermc/google-agones-sdk?projectKey=shulker).

If you believe Shulker (the operator) could help you doing that for you, do not hesitate, I'm open to the discussion and if it fits the project I would love to implement it.

**Reply by jeremylvln** [reply]
That being said, I think the behavior could be modified so:
- the Agent plugin will set the server as `Allocated` when at least one player is connected
- and this behavior can be disabled with simple config to give to hand over to any custom logic

Let me know if it would fit your needs. That can very easily be done.

**Reply by Pyjou** [reply]
I see. You right, I think it's a nice add to Shulker.
I think the main interest in shulker is the ephemeral aspect of servers, so clearer lifecycle management would be a good addition in any case and my opinion. If you need help with implementation, let me know :)

**Reply by jeremylvln** [reply, marked as answer]
I did an implementation (still untested but will do shortly). You will find the corresponding issue here: https://github.com/jeremylvln/Shulker/issues/355 (and the ongoing implementation [here](https://github.com/jeremylvln/Shulker/pull/356))

**Reply by Pyjou** [reply]
Perfect I will be testing the feature later this week. Thank you :)
